lib/xdr/types/array.ex

```diff
@@ -34,31 +36,46 @@ defmodule XDR.Type.Array do
     def build_value!(%{data_type: data_type, length: length, type_name: name} = type, raw_values)
         when is_list(raw_values) do
       unless length(raw_values) == length do
-        raise XDR.Error,
+        raise Error,
           message: "Wrong length, expected #{length} values",
           type: name,
           data: raw_values
       end
 
-      values = Enum.map(raw_values, fn value -> XDR.build_value!(data_type, value) end)
+      values =
+        raw_values
+        |> Enum.with_index()
+        |> Enum.map(fn {value, index} ->
+          Error.wrap_call(:build_value!, [data_type, value], index)
+        end)
+
       %{type | values: values}
     end
 
     def extract_value!(%{values: values}) do
-      Enum.map(values, fn value -> XDR.Type.extract_value!(value) end)
+      values
+      |> Enum.with_index()
+      |> Enum.map(fn {value, index} ->
+        Error.wrap_call(:extract_value!, [value], index)
+      end)
     end
 
     def encode!(%{values: values}) do
       values
-      |> Enum.map(&XDR.Type.encode!/1)
+      |> Enum.with_index()
+      |> Enum.map(fn {value, index} ->
+        Error.wrap_call(:encode!, [value], index)
+      end)
       |> Enum.join()
     end
 
     def decode!(%{length: length, data_type: data_type} = type, encoding) do
       {reversed_values, rest} =
-        1..length
-        |> Enum.reduce({[], encoding}, fn _, {vals, prev_rest} ->
-          {current_value, next_rest} = XDR.Type.decode!(data_type, prev_rest)
+        0..(length - 1)
+        |> Enum.reduce({[], encoding}, fn index, {vals, prev_rest} ->
```

Comment:
We might avoid the pipe operator when it's used only once.

**_Reference:_**
- https://github.com/lexmag/elixir-style-guide#needless-pipeline
- https://github.com/christopheradams/elixir_style_guide#avoid-single-pipelines